### PR TITLE
Adds support for PackageSaveMode

### PR DESCRIPTION
**Pull request by deepakaravindr**
Fixes https://github.com/NuGet/Home/issues/77

@yishaigalatzer @emgarten @zhili1208

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs [hunk 1]
```diff
@@ -17,10 +16,13 @@ public class PackageExtractionContext
         /// </summary>
         public bool UseLegacyPackageInstallPath { get; set; }
 
+        public PackageSaveModes PackageSaveMode { get; set; }
+
         public PackageExtractionContext()
         {
-            CopySatelliteFiles = false;
+            CopySatelliteFiles = true;
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs]: how is this change related?

Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs]: Actually I see below, but this can break existing codepaths, instead you should set it up when the context is null or pass it as constructor parameter.

Also note that you can in C#6 just do `{ get; set; } = value`

File: src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs [hunk 2]
```diff
@@ -154,7 +161,7 @@ public static class PackageExtractor
             }
 
             // Now, copy satellite files unless requested to not copy them
-            if (packageExtractionContext == null || packageExtractionContext.CopySatelliteFiles)
+            if ( packageExtractionContext.CopySatelliteFiles)
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs]: format

File: src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs [hunk 3]
```diff
@@ -89,8 +91,13 @@ public static class PackageExtractor
                 throw new ArgumentNullException(nameof(packagePathResolver));
             }
 
-            // TODO: Need to handle PackageSaveMode
-            // TODO: Support overwriting files also?
+            if (packageExtractionContext == null)
+            {
+                packageExtractionContext = new PackageExtractionContext();
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs]: this is where things can get pretty flaky, make it an explicit decision here.

**Comment by yishaigalatzer** [comment]
:watch:

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs (hunk 1 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs]: I will use C#6 style code. It was a mistake that the value for CopySatelliteFiles was false by default. It should be true by default. I have added tests to test that existing codepaths work properly.

File: src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs (hunk 3 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs]: Perhaps, I can change the callers to pass in a non-null PackageExtractionContext always.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs]: ok, is there a bug this fixes?

**Comment by deepakaravindr** [comment]
Incorporated the comments. Running all the tests once again. Currently, the CI has 1 test failure, same failure was observed on my box as well. Will look into it

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs [hunk 4]
```diff
@@ -17,11 +16,9 @@ public class PackageExtractionContext
...
         public bool UseLegacyPackageInstallPath { get; set; }
 
-        public PackageExtractionContext()
-        {
-            CopySatelliteFiles = false;
-            UseLegacyPackageInstallPath = false;
-        }
+        public PackageSaveModes PackageSaveMode { get; set; } = PackageSaveModes.Nupkg;
+
+        public PackageExtractionContext() { }
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs:21]: drop the constructor it is unnecessary

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs [hunk 5]
```diff
@@ -109,12 +103,12 @@ public override Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(Ca
                     await PackageExtractor.ExtractPackageAsync(
                         downloadResourceResult.PackageStream,
                         PackagePathResolver,
-                        nuGetProjectContext.PackageExtractionContext,
-                        PackageSaveMode,
+                        nuGetProjectContext.PackageExtractionContext ?? new PackageExtractionContext(),
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:106]: why wouldn't the project context have a default packageextractioncontext? Then all the ?? is not necessary. You can even go further and make it a get only property (or throw if it is set to null)

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs [hunk 6]
```diff
@@ -109,12 +103,12 @@ public override Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(Ca
...
                         token));
             }
 
-            if (PackageSaveMode.HasFlag(PackageSaveModes.Nupkg))
+            var packageSaveMode = GetPackageSaveMode(nuGetProjectContext);
+            if (packageSaveMode.HasFlag(PackageSaveModes.Nupkg))
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:111]: I know this is not your change, but I'd like to understand why we use flags enum here, where this is representing a mode

**Comment by deepakaravindr** [comment]
@zarenner, After comments are incorporated, this PR on PackageSaveMode will be merged.

**Comment by emgarten** [comment]
:shipit:

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs (hunk 4 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContext.cs:21]: Done

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs (hunk 5 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:106]: @yishaigalatzer,

I remember why I did this. INuGetProjectContext is an interface. And, setting the default value for PackageExtractionContext needs to happen for every implementation of INuGetProjectContext. Note that there multiple properties of INuGetProjectContext which are null by default. However, we have ensured that the method ExtractPackageAsync will not accept null values for PackageExtractionContext

Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:106]: :shipit:

Sent from my Windows Phone

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs (hunk 6 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:111]: As discussed offline, since there is a mode where we could extract both nuspec and nupkg, it has been flags enum.

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs (hunk 5 above)
Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:106]: Thanks! I have tracked your comment by filing a new issue: https://github.com/NuGet/Home/issues/1901

**Comment by deepakaravindr** [comment]
Merged with https://github.com/NuGet/NuGet.Client/commit/072d625d866ecf127d40894e76d38fed302b34bf and https://github.com/NuGet/NuGet.Client/commit/74fc807ddd9bb8f06880345d6704aee1607868cf
